docs/content/docs/concepts/Breaking Changes and Supported Patterns.md

```diff
@@ -0,0 +1,293 @@
+However adding the optional `override` causes the output array to be `[1, 2, 3]` which would almost certainly violate the documented behavior from before the extra argument was added. This can be avoided by doing any of the following:
+
+1. When passing a function in a way that could implicitly convert its type, wrap it in a lambda (in the example, replace `processItem` at the end with `(s) => processItem(s)`).
+2. Avoid casting functions to have more arguments. In the example replace `(item: string) => string)` with `(item: string, index: number) => string)`
+3. Don't have optional arguments.
+4. Pick one of the above for every function type usage.
+
+The first two approaches are hard to audit/enforce but allows the package to add optional arguments to functions.
+They also make adding extra arguments to callback function types an incompatible change.
+
+This makes the most pragmatic solution to do all of:
+
+1. Try and do #2 above (avoid casting functions to have more arguments). The package must do this everywhere its relevant and technically only User Code code that follows this will be considered a supported pattern.
+2. Since we are relying on avoiding implicit casting away of arguments for correctness, adding new arguments to callback functions needs to be considered an Incompatible change, even though it almost always compiles and usually works.
+3. Prefer patterns that avoid optional arguments since enforcing these rules is hard and unlikely to be done robustly without compiler assistance.
+
+Some alternatives to using optional arguments:
+
+1. The "keyword arguments" pattern, where an object is provided with named optional fields instead of optional arguments.
+   This also solves the issue of order dependency of optional arguments and is great when several optional arguments are needed.
+   The type of the object used can be an interface with the "in" variance as described in "Version Type Variance" above to allow adding members.
+2. Add alternative functions that take the extra arguments.
+   Can be combined with the above to future proof it for when even more arguments are added.
+3. Make the existing arguments more flexible.
+   For example if the function takes one string, you can replace that argument with a string or configuration object that contains the string and extra optional arguments.
+
+## Module name collisions
+
+User Code which could be broken by a member being added to an module in the Package should not be supported.
+This specifically means User Code that does things like:
+
+```typescript
+export * from "the-package";
+export { nameThatCouldCollide };
+```
+
+Is not supported.
+
+## Imports for other than the root
+
+User code which imports any path not at the root of the package should be unsupported (unless the package explicitly documents otherwise).
+This allows refactoring the package's internals.
+
+## Undocumented behavior
```

Comment:
Really love these smaller, more-concrete rules - they feel very actionable and easy to understand both for contributors and users.

Maybe also consider:

- Contributors should prefer to use the most private scope possible for all members
- Contributors should prefer to make public members `readonly`
- Contributors should use `@sealed` in any case where class extension is not explicitly a goal
- Contributors should seek to avoid class extension wherever possible, both internally and on the public interface
- Contributors should use `@internal` for any package member that isn't intended to be directly used by a customer

Reply:
If we adopt a policy like this I think we want to have a few separate documents:
1. Definition of what the supported patterns are (Source of truth for what we consider compatible)
2. Best practices for authors of "the package" designed to minimize incompatible changes as defined by the supported patterns document. Includes API extractor setup, type tests, annotations, design patterns, all the things you listed etc.
3. Set of best practices for users (referencing the supported patterns), which can include things like how to help ensure the code meets the supported pattenrs, setting up tools we provide (like a linter rule), design patterns for writing supported user code etc.. Note since we consume our own packages, this also applies to FF as well.

This document is mainly focused on `#1` for now.
I think your suggestions would be great in `#2`

Even some of my own suggestion below shouldn't be in this shouldn't be by this argument. I may have to refactor things clean this up.

This makes me wonder perhaps structuring this in-terms of Package invariants instead of Supported Patterns would be more intuitive. We still get three sections:
1. Package Invariants
2. How to implement the package providing useful invariants
3. How to consume the package such that the user code won't break as long as the invariants hold.

Such a construction makes it easier to give ourselves wiggle room for some TypeScript edge cases (ex: we can promise types will be assignable, but not that you won't get build errors due to compiler errors about no longer used members of "in" types when constructing typed literals and strict flags enabled). I'll have to think about this more.